### ISSUE TITLE
test-isolation-enforcement: MODULE_* patterns match at any indentation, the 70% setup rule is Jest-shaped, and a comment counts as a resource

### ISSUE THREAD

**Issue by ruleofcodedev** [Rule-Of-Code/RoC#124]
**Version:** 7.22.0
**Law:** `test-isolation-enforcement` (severity: error — it fails the build)

First: thank you for the `ignores.tests` fix. This law and `test-documentation-requirements` went from a silent 100/100 to a real verdict, which is exactly what should have happened. `test-documentation-requirements` found two things that were genuinely ours and they are now paid — one file with no documentation, one test case with an unclear description.

This one reports three violations, and measuring each against our suite, none of the three is real.

```
1. Test Isolation Enforcement [3 violations] (Score: 25/100)
   • Tests share mutable state that can cause interference
   • Tests missing proper isolation in setup/teardown
   • External resources not properly isolated between tests
```

## 1. "Shared mutable state" — 6 files flagged, 0 at module scope and mutable

```js
SHARED_MUTABLE_PATTERNS = {
  MODULE_LET:            /^\s*let\s+\w+(?:\s*:\s*[^=]+)?\s*=(?!.*beforeEach)/m,
  MODULE_MUTABLE_OBJECT: /^\s*const\s+\w+\s*=\s*\{/m,
  MODULE_MUTABLE_ARRAY:  /^\s*const\s+\w+\s*=\s*\[/m,
}
```

The names say MODULE, but `^\s*` with `/m` matches at any indentation, so a declaration inside a test body counts as module state. Every match in our suite, with its indentation:

```
docs.spec.ts:76          INDENTED   const shown = ['laws.minLawsChecked', …]
home-layout.spec.ts:138  INDENTED   const sections = [...document.querySelectorAll('section')]
hydration.spec.ts:29     INDENTED   const pages = ['index.html', join('laws','index.html')]
journey.perf.spec.ts:117 INDENTED   const heaviest = [...scripts]
mobile.spec.ts:33        INDENTED   let parent = el.parentElement;
mobile.spec.ts:22        module     const PHONE = { width: 390, height: 844 };
theme.spec.ts:13         INDENTED   const documentElement = {
```

Five of the seven are function-local, and `let parent = el.parentElement` is the cursor of a `while` loop walking up the DOM — as local as a variable gets. `const documentElement = {` is inside a factory that builds a fresh fake `Document` on every call, which is isolation, not the absence of it.

The one match at module scope is `const PHONE = { width: 390, height: 844 }` — a viewport constant that is never assigned to. Extracting those two numbers into a named constant is what `magic-number-prevention` asks for in the same audit.

There is also no mutation analysis: `const` is an immutable binding, and an object literal that is never written to cannot leak between tests. Anchoring to column 0 would remove five of these; requiring evidence of a later write would remove the sixth.

## 2. "Missing proper isolation in setup/teardown" — a Jest shape applied to Playwright

`hasProperSetupTeardown` requires `beforeEach`, `afterEach`, `TestBed.resetTestingModule`, or a mock reset, in at least 70% of test files. Ours:

```
17 of 29 files have none  ->  41.4%, below the 70% threshold
```

Almost all 17 are Playwright specs, where isolation is the fixture rather than a hook: `test('…', async ({ page }) => …)` gives every test a fresh `BrowserContext` and `Page`. Adding empty `beforeEach` blocks to reach 70% would be ceremony that changes nothing about whether the tests are isolated — and this repository has a written rule against exactly that kind of edit.

The Jest specs among the 17 have no state to reset: one tests a pure function against a fake `Document` built per call, another reads two files and compares strings. The one Jest spec that does touch `document.documentElement` is `app.spec.ts`, and it has both `beforeEach` and `TestBed.resetTestingModule`.

A Playwright file is recognisable — `from '@playwright/test'`, and tests declared with a destructured fixture argument. Treating that as isolation, or excluding those files from this particular sub-check, would make the 70% mean something.

## 3. "External resources not properly isolated" — one match, inside a comment

```
apps/web/src/app/app.spec.ts:138
 * `ThemeStorage.read()` wraps localStorage in try/catch and returns null on failure
```

That is the only occurrence of `localStorage` in the file — a line of prose inside a block comment explaining why a test was deleted. The scan runs over raw content, so a comment that *describes* a resource is read as *using* one. Same family as #105, except there it was string literals; the comment-stripping the magic-number analyzer already does would fix this one.

## What we are doing meanwhile

Moving this law to `warning` with the reason written into our config and this issue number beside it, per our own ladder: a law is `error`, or it is `notApplicable` with a written reason, or it is a checker defect that is reported and parked with the issue number. It is parked, not waived — and if any of the three turns out to be real after you look, we will pay it.

**Comment by ruleofcodedev** [comment]
Fixed in #131, merged to `develop`. Ships in the next release.

All three sub-checks. Measured on your tree: shared state `true` → `false`, isolation ratio `false` → `true` (21/30 = 70.0%).

Resource isolation stays `false`, and that one is correct: `apps/web-e2e/src/theme.spec.ts` calls `page.addInitScript(() => localStorage.setItem('roc-theme', 'banana'))` with no cleanup. Real use of a real resource, not prose about one — the comment strip did not sweep it away. I checked that separately: 3 files matched a resource pattern in their raw text, 2 in their code, so exactly one was a comment, and it was the one you reported.

Two notes on how it was fixed:

- `CodeText.stripComments` already existed in this repo. The law simply was not calling it — same shape as the config-path defects in 7.21.0.
- Swept: `unit-test-data-management` carries a second copy of all three checks. Both now call the shared `FixtureIsolation`, so a project cannot pass one isolation law and fail its identical twin.

Two existing tests asserted the defect (`const config = {` with nothing writing to it counted as shared state). Changed deliberately, with the reason recorded in the test: `magic-number-prevention` demands exactly such a constant in the same audit.

**Comment by ruleofcodedev** [comment]
Re-tested against 7.23.0. Two of the three are fixed and the law went from 3 findings to 1 here — thank you.

```
before 7.23.0   Test Isolation Enforcement [3 violations] (25/100)
after           Test Isolation Enforcement [1 violations] (80/100)
```

Gone: the `MODULE_*` patterns no longer match a declaration inside a test body, and the 70% setup/teardown rule no longer fires on a Playwright suite.

## What is left, and part of it is now genuinely ours

`External resources not properly isolated between tests` still fires, and this time there are two sources, not one:

```
theme.spec.ts:119   CODE      await page.addInitScript(() =>
                              localStorage.setItem('roc-theme', 'banana'))
app.spec.ts:138     COMMENT    * `ThemeStorage.read()` wraps localStorage in
                              try/catch and returns null on failure
```

The first is **ours and new** — a test added since the last report, which really does write to storage. The law was not right about us before and is now; I mention it because it changes what I am asking for.

The second is the comment one from the original report, unchanged. It is the only occurrence of the word in that file: prose inside a block comment, explaining why a test was deleted, read as a test *using* a resource.

## The fix you already have

`AngularCleanupConfiguration.stripComments` exists in the tool, and several checkers call it — `magic-number-prevention`, `centralized-logging`, `no-double-cast`, `no-explicit-any-non-null`. `isolation-analyzer.service.js` does not, and neither does the `usesExternalResources` path it feeds.

This is the same shape as #105: the helper was already written, the law simply was not calling it.

## The half I am not asking you to change

For the Playwright case, "not properly isolated" has no fix that means anything: every test gets a fresh `BrowserContext`, so storage written in one is unreachable from the next. The isolation is the fixture. Adding an `afterEach` that clears a key nothing else can see would satisfy the check and change nothing about the product — the box-ticking our own policy forbids.

So I would rather leave that one visible than pay it dishonestly, and I have written that into our config beside the law. If you would rather the check recognised a Playwright fixture the way the setup/teardown one now does, that is your call, not something I want to press.
